src/buffer.rs

```diff
@@ -394,6 +415,109 @@ impl<'a> PartialEq for Cursor<'a> {
     }
 }
 
+/// Comparison of `Cursor`s is not as cheap as other cursor operations. This is
+/// because it requires walking the token stream to handle `None`-delimited groups.
+impl<'a> PartialOrd for Cursor<'a> {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        // If we don't have the same scope end, the cursors are incomparable.
+        if self.scope != other.scope {
+            return None;
+        }
```

Comment:
Oof, I didn't realize this is hard to handle. I wonder if this makes the whole impl fruitless for your use case. For example your "rooted, group before" or "different group" test cases, but with parens instead of None groups:

```py
# rooted, group before
∘ ( ∘ ● ∘ ) ∘ ● ∘

# different group
∘ ( ∘ ● ∘ ) ∘ ( ∘ ● ∘ ) ∘
```

One would probably require the first black dot to compare less than the second black dot, but with this impl `a < b` and `b < a` would both be false.

One would need to know to write `a < b && !(b < a)` in order to get an actual meaningful comparison, or `a.partial_cmp(b)` directly.

Can we just safely keep going with both cursors regardless of scope? The very very outermost `Entry::End` contains ptr::null(), so we can look for that, instead of paying attention to `scope` at all. Basically my mental model is this impl shouldn't care at all about any distinction between non-None groups, explicitly entered None groups, and implicitly entered None groups. The only point of `scope` is distinguishing the implicitly entered None case from the other two (if I'm recalling correctly).

Reply:
Okay I realize the objection: that's inconsistent with PartialEq. If we ignore scope, PartialEq could say cursors are not equal while PartialOrd could say they are equal (same ptr, different scope).

I wonder what could possibly go wrong if PartialEq looked at ptr only.

Reply:
Okay the existing PartialEq impl for Cursor is probably just wrong and bad. The only place in our codebase that uses it assumes it works the other way, and crashes with the current impl.

```rust
// [dependencies]
// syn = { version = "1", features = ["full", "extra-traits"] }
// proc-macro2 = "1"
// quote = "1"

use proc_macro2::{Delimiter, Group};
use quote::quote;

fn main() {
    // Okay. Rustc allows top-level `static` with no value syntactically, but
    // not semantically. Syn parses as Item::Verbatim.
    let tokens = quote!(pub static FOO: usize; pub static BAR: usize;);
    let file = syn::parse2::<syn::File>(tokens).unwrap();
    println!("{:#?}", file);

    // Okay.
    let inner = Group::new(Delimiter::None, quote!(static FOO: usize = 0; pub static BAR: usize = 0));
    let tokens = quote!(pub #inner;);
    let file = syn::parse2::<syn::File>(tokens).unwrap();
    println!("{:#?}", file);

    // Parser crash.
    let inner = Group::new(Delimiter::None, quote!(static FOO: usize; pub static BAR: usize));
    let tokens = quote!(pub #inner;);
    let file = syn::parse2::<syn::File>(tokens).unwrap();
    println!("{:#?}", file);
}
```

Reply:
Should I "fix" equality to only care about `ptr` as well? This is safe, comparing cursor positions ignoring "partial eof" does seem generally more useful (and certainly is for the "which cursor is further along" query[^1]).

[^1]: Although it turns out my use case doesn't have that cursor; a parse stream's cursor is IIUC in an unspecified state after parse error and [in practice](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=96011ab853b9c9744d37ea485141ccb9) is after the block that the error occurred in, not at the position of the error. Providing the cursor with the error is impossible since `Error: 'static` (and that's a good thing; much better than the relatively minor annoyance of ).

If we do decide to do comparison ignoring "partial eof," I think the routine should be

- If `self.ptr == other.ptr`, return `Equal`.
- Walk `self.ptr` to `self.scope`, recording traversed ranges.
  - If `other.ptr` is in these ranges, return `Less`. (This step is purely an early-out optimization.)
- Walk `other.ptr` to `other.scope`, recording traversed ranges (iff `self.scope != other.scope`).
  - If we enter `self`'s traversed ranges, return `self_range.start <=> other_range.start`.
  - If `self.scope == other.scope`, `unreachable!()`.
- Walk `self.scope` to `End(nullptr)`, recording traversed ranges (iff `other.scope != End(nullptr)`).
  - If we enter `other`'s traversed ranges, return `self_range.start <=> other_range.start`.
- Walk `other.scope` to `End(nullptr)`.
  - If we enter `self`'s traversed ranges, return `self_range.start <=> other_range.start`.
- return `Incomparable`.

since compared cursors being in the same scope is still a common case worth optimizing for, I think.

This is IIUC $O\left( d_{self} b_{self}  + d_{other} b_{other} + d_{self} d_{other} \right)$ time and $O\left( d_{self} + d_{other}' \right)$ space (where $d$ is cursor depth, $d'$ is cursor depth-to-scope (negligible), and $b$ is block length), and I think optimal big-$O$ only achieves that complexity with $d' = 0$ (by walking `self` all the way first).

More in the style of big-$O$, that's $O(d^2)$ time and $O(d)$ space. The polynomial is unfortunate but _required_ without a tree restructuring[^2] or storing sufficient-for-comparison information on the cursors[^3]. The tree restructuring is a big ask, and storing more on `Cursor` is penalizing normal `LL(3)` consumers for niche consumers who want to compare cursors.

[^2]: e.g. a flat representation that makes cursor comparison purely `self.ptr <=> other.ptr`.

[^3]: e.g. an index of tokens yielded since the beginning of the root buffer, and a pointer to the root buffer to check they're comparable if you want to not just give consistent-but-meaningless results for comparing cursors in different root `TokenBuffer`s.

Just counting tokens-to-end is $O(2n)$ time, $O(1)$ space; $n$ is the flattened length of the token stream. Since $d \approx \log n$, the comparison walk is $\approx O(\log^2 n)$ and better than the trivial tokens-to-end impl in nearly all[^4] cases.

[^4]: The exceptions also being somewhat mitigated by early-exit, since when $d \not\ll n$, much of the nesting is likely shared. Exceptions are constructed and not likely in practice.

The interesting thing to note, though, is that what is _actually_ wanted is `[Cursor]::sort`, *not* `Cursor::partial_ord`, and comparing for `sort` can skip recalculating the walk-to-`End(nullptr)` list for each cursor on each sort. Though IIUC this won't show up in big-$O$, it could be a significant improvement in practice.

Reply:
> Should I "fix" equality to only care about ptr as well?

Yes I think we should change this. I would be shocked if anyone had found a use for the current behavior, and I would be shocked if changing it wouldn't *fix* a significant fraction of places the impl is used, much like it fixes our own buggy use.